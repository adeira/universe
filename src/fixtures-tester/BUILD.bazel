load("//:bazel/macros/build_yarn_workspace.bzl", "build_yarn_workspace")
load("//:bazel/macros/test_yarn_workspace.bzl", "test_yarn_workspace")

build_yarn_workspace(
    name = "fixtures-tester",
    package_name = "@adeira/fixtures-tester",
)

test_yarn_workspace(
    name = "test",
)
