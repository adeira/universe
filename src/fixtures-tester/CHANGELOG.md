# Unreleased

# 1.1.0

- New function `verifyTestsFromFixtures` added - it has similar interface as `generateTestsFromFixtures` except it verifies that the operation result is `true` instead of creating Jest snapshots. This is useful when you want to run some tests on your fixtures files but there is not much to snapshot.

# 1.0.1

Note for all Flow users: all projects in [`adeira/universe`](https://github.com/adeira/universe) now use implicit exact Flow types (`{}` for strict objects and `{ ... }` for open objects, syntax `{||}` is deprecated). We do not expect any issues as long as you are using `exact_by_default=true` Flow option.

- Fix failure on unexpected subfolders, see: https://github.com/adeira/universe/pull/2257

# 1.0.0

Initial release.
