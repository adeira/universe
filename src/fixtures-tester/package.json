{
  "name": "@adeira/fixtures-tester",
  "description": "Utility to ease testing with fixtures.",
  "homepage": "https://github.com/adeira/fixtures-tester",
  "bugs": "https://github.com/adeira/universe/issues",
  "repository": {
    "type": "git",
    "url": "git@github.com:adeira/universe.git",
    "directory": "src/fixtures-tester"
  },
  "license": "MIT",
  "private": false,
  "version": "1.1.1",
  "main": "./src/index.js",
  "type": "commonjs",
  "sideEffects": false,
  "dependencies": {
    "@adeira/js": "^2.1.1",
    "@babel/runtime": "^7.16.7"
  }
}
