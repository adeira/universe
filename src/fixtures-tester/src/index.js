// @flow

export { default as generateTestsFromFixtures } from './generateTestsFromFixtures';
export { default as verifyTestsFromFixtures } from './verifyTestsFromFixtures';
