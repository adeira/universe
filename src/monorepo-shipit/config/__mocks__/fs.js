// @flow

const fs: any = jest.createMockFromModule('fs');

fs.existsSync = (path) => path !== '/unknown_path';

export default fs;
