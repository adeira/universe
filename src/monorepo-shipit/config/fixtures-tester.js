// @flow strict

import type { ConfigType } from '../ConfigType.flow';

module.exports = ({
  getStaticConfig() {
    return {
      repository: 'git@github.com:adeira/fixtures-tester.git',
    };
  },
  getPathMappings() {
    return new Map([['src/fixtures-tester/', '']]);
  },
}: ConfigType);
