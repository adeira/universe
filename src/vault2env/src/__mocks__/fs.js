// @flow

const fs: any = jest.createMockFromModule('fs');

let mockFiles = Object.create(null);
fs.__setMockFiles = function (newMockFiles) {
  mockFiles = newMockFiles;
};

fs.existsSync = function (filePath) {
  return mockFiles[filePath] !== undefined;
};

fs.writeFileSync = function (filePath, content) {
  mockFiles[filePath] = content;
};

fs.readFileSync = function (filePath) {
  return mockFiles[filePath];
};

module.exports = fs;
