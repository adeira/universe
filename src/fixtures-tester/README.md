Install (via Yarn or NPM):

```text
yarn add @adeira/fixtures-tester
```

## `generateTestsFromFixtures`

This function loads all the fixture files (their file content) from your folder and snapshots the result of your callback operation. This is handy when you have several example files, you want to perform some operation on top of them and snapshot the results.

```js
import { generateTestsFromFixtures } from '@adeira/fixtures-tester';

generateTestsFromFixtures(
  `${__dirname}/__fixtures__`,
  (input: string): string => {
    return doSomethingWithTheInput(input); // this result will be saved in a Jest snapshot
  },
  // optionally a snapshot name
);
```

There are two additional modes which allow you to control the fixtures behavior more precisely. There are based on the fixture name:

- `simpleExample.graphql` - fixture will snapshot as expected
- `simpleExample.only.graphql` - only these snapshots will run (for debug only)
- `simpleExample.error.graphql` - it's expected that fixture will throw an error (expected to fail tests)

The error file convention is required to prevent unintentional snapshots of exceptions.

## `verifyTestsFromFixtures`

This function simply verifies that the callback returns `true` but it doesn't snapshot anything:

```js
import { verifyTestsFromFixtures } from '@adeira/fixtures-tester';

verifyTestsFromFixtures(`${__dirname}/__fixtures__`, (input: string): boolean => {
  return isInputValid(input) === true;
});
```

This is handy when the snapshot from `generateTestsFromFixtures` would be meaningless but you still want to check all available fixtures.
