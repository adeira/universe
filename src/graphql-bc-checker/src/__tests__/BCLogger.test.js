// @flow

import fs from 'fs';
import path from 'path';
import stripAnsi from 'strip-ansi';
import { generateTestsFromFixtures } from '@adeira/fixtures-tester';

import { buildBreakingChangesBlock } from '../BCLogger';
import testBackwardCompatibility from '../index';

function operation(newBreakingChanges) {
  return (input) => buildBreakingChangesBlock(input, newBreakingChanges);
}

generateTestsFromFixtures(path.join(__dirname, '__fixtures__'), operation());
generateTestsFromFixtures(
  path.join(__dirname, '__fixtures__'),
  operation([
    {
      type: 'append',
      description: 'this',
    },
    {
      type: 'and append',
      description: 'also this',
    },
  ]),
);

function stringifyMockCalls(calls: Array<string>): Array<string> {
  return calls.map((call) => {
    return stripAnsi(call.toString().trim());
  });
}

it('prints success message when there are no breaking changes', () => {
  const consoleSpy = jest.spyOn(console, 'log').mockImplementation(() => {});

  testBackwardCompatibility({
    allowBreakingChanges: false,
    snapshotLocation: path.join(__dirname, 'testSchemaSnapshot.graphql'),
    schema: require('./testSchema').validSchema,
  });

  expect(stringifyMockCalls(consoleSpy.mock.calls)).toMatchInlineSnapshot(`
    Array [
      "Testing: testSchemaSnapshot.graphql",
      "Congratulations! NO BREAKING CHANGES or OUTDATED SCHEMA. Good job!",
    ]
  `);

  jest.restoreAllMocks();
});

it('prints error messages when there are breaking changes', () => {
  const consoleLogSpy = jest.spyOn(console, 'log').mockImplementation(() => {});
  const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
  const processSpy = jest.spyOn(process, 'exit').mockImplementation((code) => {
    throw new Error(`process.exit(${code}) was called`);
  });

  expect(() =>
    testBackwardCompatibility({
      allowBreakingChanges: false,
      snapshotLocation: path.join(__dirname, 'testSchemaSnapshot.graphql'),
      schema: require('./testSchema').breakingSchema,
    }),
  ).toThrow('process.exit(1) was called');

  expect(stringifyMockCalls(consoleErrorSpy.mock.calls)).toMatchInlineSnapshot(`
    Array [
      "You introduced breaking changes into the public GraphQL schema. This change may or may not be intentional. These breaking changes may break some clients consuming our public API. Please try to find a way how to avoid breaking changes and try it again. Here is list of all breaking changes:",
      "FIELD_REMOVED - RootQuery.test was removed.",
    ]
  `);
  expect(stringifyMockCalls(consoleLogSpy.mock.calls)).toMatchInlineSnapshot(`
    Array [
      "Testing: testSchemaSnapshot.graphql",
      "Tips how to avoid breaking changes:

    - field removal/modification (introduce new field and only deprecate the old one)
    - type removal/modification (just deprecate it and leave it there)
    - removal from enum/union (introduce new enum/union)
    - arguments removal/modification (introduce new query or graph node)
    - change non-nullable -> nullable (just don't do it or introduce new field)
    - change of default argument value (don't or introduce new argument/query)",
    ]
  `);
  expect(processSpy).toHaveBeenCalledWith(1);

  jest.restoreAllMocks();
});

it('prints error messages when the schema is manually polluted', () => {
  const consoleSpy = jest.spyOn(console, 'log').mockImplementation(() => {});
  const processSpy = jest.spyOn(process, 'exit').mockImplementation((code) => {
    throw new Error(`process.exit(${code}) was called`);
  });

  expect(() =>
    testBackwardCompatibility({
      allowBreakingChanges: false,
      snapshotLocation: path.join(__dirname, 'testSchemaSnapshotPolluted.graphql'),
      schema: require('./testSchema').validSchema,
    }),
  ).toThrow('process.exit(1) was called');
  expect(stringifyMockCalls(consoleSpy.mock.calls)).toMatchInlineSnapshot(`
    Array [
      "Testing: testSchemaSnapshotPolluted.graphql",
      "Manual changes of GraphQL snapshot detected. Please do not update GraphQL snapshot manually. This file is being autogenerated.",
    ]
  `);
  expect(processSpy).toHaveBeenCalledWith(1);

  jest.restoreAllMocks();
});

it('prints warning and updates the schema when backward compatible changes detected', () => {
  const fsSpy = jest.spyOn(fs, 'writeFileSync').mockImplementation(() => {});
  const consoleSpy = jest.spyOn(console, 'log').mockImplementation(() => {});
  const processSpy = jest.spyOn(process, 'exit').mockImplementation((code) => {
    throw new Error(`process.exit(${code}) was called`);
  });

  expect(() =>
    testBackwardCompatibility({
      allowBreakingChanges: false,
      snapshotLocation: path.join(__dirname, 'testSchemaSnapshot.graphql'),
      schema: require('./testSchema').compatibleSchema,
    }),
  ).toThrow('process.exit(1) was called');

  expect(fsSpy.mock.calls[0][1]).toMatchSnapshot();
  expect(stringifyMockCalls(consoleSpy.mock.calls)).toMatchInlineSnapshot(`
    Array [
      "Testing: testSchemaSnapshot.graphql",
      "GraphQL schema snapshot IS OUTDATED! (updating automatically)",
      "Snapshot of the GraphQL schema successfully created! In case you see this message in CI you have to run locally command \`yarn test-bc\` and commit the changes.",
    ]
  `);
  expect(processSpy).toHaveBeenCalledWith(1);

  jest.restoreAllMocks();
});
