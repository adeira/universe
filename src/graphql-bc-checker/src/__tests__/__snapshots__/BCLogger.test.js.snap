// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches expected output: empty-block.txt 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
# </BREAKING-CHANGES-LOG>

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
# </BREAKING-CHANGES-LOG>
`;

exports[`matches expected output: empty-block.txt 2`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
# </BREAKING-CHANGES-LOG>

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  append: this
#  and append: also this
# </BREAKING-CHANGES-LOG>
`;

exports[`matches expected output: full-block.txt 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  FIELD_REMOVED: Booking.unstable_baggage was removed.
#  FIELD_REMOVED: BookingOneWay.unstable_baggage was removed.
#  FIELD_REMOVED: BookingInterface.unstable_baggage was removed.
#  FIELD_REMOVED: BookingReturn.unstable_baggage was removed.
#  FIELD_REMOVED: BookingMulticity.unstable_baggage was removed.
# </BREAKING-CHANGES-LOG>

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  FIELD_REMOVED: Booking.unstable_baggage was removed.
#  FIELD_REMOVED: BookingOneWay.unstable_baggage was removed.
#  FIELD_REMOVED: BookingInterface.unstable_baggage was removed.
#  FIELD_REMOVED: BookingReturn.unstable_baggage was removed.
#  FIELD_REMOVED: BookingMulticity.unstable_baggage was removed.
# </BREAKING-CHANGES-LOG>
`;

exports[`matches expected output: full-block.txt 2`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  FIELD_REMOVED: Booking.unstable_baggage was removed.
#  FIELD_REMOVED: BookingOneWay.unstable_baggage was removed.
#  FIELD_REMOVED: BookingInterface.unstable_baggage was removed.
#  FIELD_REMOVED: BookingReturn.unstable_baggage was removed.
#  FIELD_REMOVED: BookingMulticity.unstable_baggage was removed.
# </BREAKING-CHANGES-LOG>

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  FIELD_REMOVED: Booking.unstable_baggage was removed.
#  FIELD_REMOVED: BookingOneWay.unstable_baggage was removed.
#  FIELD_REMOVED: BookingInterface.unstable_baggage was removed.
#  FIELD_REMOVED: BookingReturn.unstable_baggage was removed.
#  FIELD_REMOVED: BookingMulticity.unstable_baggage was removed.
#  append: this
#  and append: also this
# </BREAKING-CHANGES-LOG>
`;

exports[`matches expected output: missing-block.txt 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
# </BREAKING-CHANGES-LOG>
`;

exports[`matches expected output: missing-block.txt 2`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  append: this
#  and append: also this
# </BREAKING-CHANGES-LOG>
`;

exports[`prints warning and updates the schema when backward compatible changes detected 1`] = `
"# @generated SignedSource<<d309d94b9c45979a5e0b9a4ddaba6722>>

# <BREAKING-CHANGES-LOG>
# </BREAKING-CHANGES-LOG>

schema {
  query: RootQuery
}

\\"\\"\\"The query root of the schema\\"\\"\\"
type RootQuery {
  test: Test
  test2: Test2
}

\\"\\"\\"A test object type\\"\\"\\"
type Test {
  id: String
}

\\"\\"\\"A test object type\\"\\"\\"
type Test2 {
  id: String
}"
`;
